File: dodam-core/src/main/java/b1nd/dodamcore/outsleeping/application/OutSleepingService.java
package b1nd.dodamcore.outsleeping.application;

import b1nd.dodamcore.member.domain.entity.Student;
import b1nd.dodamcore.outsleeping.domain.entity.OutSleeping;
import b1nd.dodamcore.outsleeping.domain.enums.OutSleepingStatus;
import b1nd.dodamcore.outsleeping.domain.exception.OutSleepingNotFoundException;
import b1nd.dodamcore.outsleeping.repository.OutSleepingRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class OutSleepingService {

    private final OutSleepingRepository repository;

    public void save(OutSleeping outSleeping) {
        repository.save(outSleeping);
    }

    public void delete(OutSleeping outSleeping) {
        repository.delete(outSleeping);
    }

    public OutSleeping getById(Long id) {
        return repository.findById(id)
                .orElseThrow(OutSleepingNotFoundException::new);
    }

    public List<OutSleeping> getValid(LocalDate now) {
        return getByDate(now).parallelStream()
                .filter(o -> OutSleepingStatus.ALLOWED.equals(o.getStatus()))
                .toList();
    }

    public List<OutSleeping> getByDate(LocalDate date) {
        return repository.findByDate(date);
    }

    public List<OutSleeping> getByStudent(Student student, LocalDate now) {
        return repository.findByStudentAndEndAtGreaterThanEqual(student, now);
    }

}

File: services/service-out-sleeping/src/main/kotlin/com/b1nd/dodamdodam/outsleeping/domain/outsleeping/service/OutSleepingService.kt
package com.b1nd.dodamdodam.outsleeping.domain.outsleeping.service

import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.entity.OutSleepingEntity
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.enumeration.OutSleepingStatus
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.exception.OutSleepingAlreadyProcessedException
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.exception.OutSleepingNotFoundException
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.exception.OutSleepingNotOwnerException
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.repository.OutSleepingRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.UUID

@Service
class OutSleepingService(
    private val outSleepingRepository: OutSleepingRepository
) {

    fun create(outSleeping: OutSleepingEntity): OutSleepingEntity =
        outSleepingRepository.save(outSleeping)

    fun getById(id: Long): OutSleepingEntity =
        outSleepingRepository.findById(id).orElseThrow { OutSleepingNotFoundException() }

    fun getByUserId(userId: UUID): List<OutSleepingEntity> =
        outSleepingRepository.findAllByUserId(userId)

    fun getByDate(date: LocalDate, pageable: Pageable): Page<OutSleepingEntity> =
        outSleepingRepository.findAllByStartAtLessThanEqualAndEndAtGreaterThanEqual(date, date, pageable)

    fun getByStatusAndDate(status: OutSleepingStatus, date: LocalDate): List<OutSleepingEntity> =
        outSleepingRepository.findAllByStatusAndStartAtLessThanEqualAndEndAtGreaterThanEqual(
            status, date, date
        )

    fun getAllowedByDate(date: LocalDate): List<OutSleepingEntity> =
        getByStatusAndDate(OutSleepingStatus.ALLOWED, date)

    fun validateOwner(outSleeping: OutSleepingEntity, userId: UUID) {
        if (outSleeping.userId != userId) {
            throw OutSleepingNotOwnerException()
        }
    }

    fun validatePending(outSleeping: OutSleepingEntity) {
        if (outSleeping.status != OutSleepingStatus.PENDING) {
            throw OutSleepingAlreadyProcessedException()
        }
    }

    fun delete(outSleeping: OutSleepingEntity) =
        outSleepingRepository.delete(outSleeping)
}
